### DISCUSSION TITLE
register_looped stops looping after some time

### DISCUSSION THREAD

**Discussion by MindOfBog** [YimMenu/YimMenu#3640] [Q&A] [answered]
Ok so I've written a little regen script here, and it works exactly how I want it to, however it seems that at random intervals this loop stops working, I have not found a way to replicate it so I can't really debug it. Is there something in the loop that could cause this? I initially thought maybe the player ped becomes null at some points and it causes a null pointer exception that crashes the loop but even after the null check it still happens...

```lua
script.register_looped("regenLoop", function ()
    local currentMillis = getMillis()
    local player = getPlayer()

    if not (player == nil) then

        local maxHealth = PED.GET_PED_MAX_HEALTH(player)
        local playerHealth = ENTITY.GET_ENTITY_HEALTH(player)
        local isAlive = not PED.IS_PED_DEAD_OR_DYING(player, true)

        if regenHP < 0 then
            regenHP = 0
        end
        
        if regenHP > maxHealth then
            regenHP = maxHealth
        end
        
        if regenArmorAmount < 0 then
            regenArmorAmount = 0
        end
        
        if regenArmorAmount > maxHealth then
            regenArmorAmount = maxHealth
        end

        if regenDelay < 0 then
            regenDelay = 0
        end

        if regenArmorDelay < 0 then
            regenArmorDelay = 0
        end

        if regen:is_enabled() and isAlive and playerHealth < maxHealth and currentMillis - hpRegenMillis >= regenDelay then
            ENTITY.SET_ENTITY_HEALTH(player, playerHealth + regenHP, player, joaat("EgoChaser"))
            hpRegenMillis = currentMillis
        end

        if regenArmor:is_enabled() and isAlive and currentMillis - armorRegenMillis >= regenArmorDelay then
            PED.ADD_ARMOUR_TO_PED(player, regenArmorAmount)
            armorRegenMillis = currentMillis
        end
    end
end)
```

Also sort of unrelated, since this is my first time scripting for gta, am I using the correct functions to set HP and Armor? or is there a better way to do it without it being detectable? My thought here is this could look like eating snacks or equipping armor.

**Comment by xesdoog** [comment, marked as answer]
You're overcomplicating it.
1. `PED.IS_PED_DEAD_OR_DYING()` is meant for NPCs, not players (it checks for a writhe task which players never go into, they just die). Just use `ENTITY.IS_ENTITY_DEAD()` or check if playerHealth > 0
2.  `getMillis()` and the health and armor regen millis and delays are all useless. You can simply declare an integer variable that takes its value from an ImGui SliderInt or InputInt and then sleep for that amount.
3. `getPlayer()` is also useless. Just use `self.get_ped()`. The devs have put so much effort into making the Lua API [easy to use](https://github.com/YimMenu/YimMenu/tree/master/docs/lua/tables).

> maybe the player ped becomes null

Null is not a Lua type, it's nil and we can't know if it's the cause of your issue since you didn't include your `getPlayer()` function.
> I can't really debug it

You can always log information to the console to check if something is working as intended or not.

>  `ENTITY.SET_ENTITY_HEALTH(player, playerHealth + regenHP, player, joaat("EgoChaser"))`

The function takes a weapon hash as the last parameter. EgoChaser is not a weapon and running a joaat hash function on it is returning this number **1333816806** which doesn't mean anything to the game.

> My thought here is this could look like eating snacks or equipping armor.

You're over-thinking, bud.

Try this example:

```lua
heal_tab = gui.add_tab("Auto Heal")
-- declare your locals
local auto_heal   = false
local regen_delay = 0
-- add your GUI
heal_tab:add_imgui(function()
  auto_heal, _ = ImGui.Checkbox("Auto Heal", auto_heal) -- Use direct ImGui calls. It's much easier to say: `if auto_heal then ... end` instead of `if auto_heal:is_enabled() then ... end`
  if auto_heal then
    ImGui.Text("Regen Delay:")
    regen_delay, _ = ImGui.SliderInt("##delay", regen_delay, 0, 1000) -- delay between 0 and 1000ms
  end
end)

-- create your thread
script.register_looped("autoheal", function(script)
  script:yield() -- yield the execution until the next frame
  if auto_heal and not ENTITY.IS_ENTITY_DEAD(self.get_ped()) then -- if the checkbox is enabled and localPlayer is not dead
    local maxHealth    = PED.GET_PED_MAX_HEALTH(self.get_ped())
    local maxArmour    = 50 -- normal armour in Online
    local playerHealth = ENTITY.GET_ENTITY_HEALTH(self.get_ped())
    local playerArmour = PED.GET_PED_ARMOUR(self.get_ped())
    if playerArmour == nil then -- when your armour is completely depleted, it will be read as nil, not 0 which will cause ADD_ARMOUR_TO_PED() to not work and throw errors. This is most probably why your script stops working (if it stops after you die and you have regenArmor enabled then this is definitely why)
      PED.SET_PED_ARMOUR(self.get_ped(), 10) -- you can set it to 50 if you want. It will give you max armour as soon as you respawn after dying.
    end
    if playerHealth < maxHealth. -- A loop repeats itself, meaning your looped script will keep checking a condition and once it's no longer true, the execution will stop. You don't have to manually check and stop it.
      ENTITY.SET_ENTITY_HEALTH(self.get_ped(), playerHealth + 1, 0, 0) -- The last 2 parameters you set in your script are useless. Setting both to 0 works just fine and it's what R* does in the game's scripts.
    end
    if playerArmour ~= nil and playerArmour < maxArmour then
      PED.ADD_ARMOUR_TO_PED(self.get_ped(), 1) 
    end
    script:sleep(regen_delay)
  end
end)
```

You control how fast you regain health and armour using the time delay.
You can further customize the regen speed by changing the amount gained. Health can be set to as little as +0.1 per tick but armour can not be set to less than +0.5 per tick, otherwise it won't do anything.

**Reply by MindOfBog** [reply]
Ah I see, thanks for the explanations, I hadn't really looked at the docs much so apologies for that, I have been browsing and searching for information on older forum posts and such to get my information.

Now the loop didn't stop working after dying so I'm still unsure if the Armor thing was the issue (it could well be), it stopped at random times, most of the times I didn't even notice long after, until I started taking damage again, so like after playing a mission or going from SP to MP. But I can't pin down what it actually is, because it works under all of these conditions when testing them.

But I will test it out again with these changes to see if it still happens.

`getMillis()` is just to get current milliseconds via `os.clock()` so that I can not only have separate variable delays for each regen but also do more stuff within the loop without everything in it being constrained by the sleep.

Thanks a bunch

**Reply by MindOfBog** [reply]
Side note: whenever I put yield into the loop, it stops working completely.

**Reply by ShinyWasabi** [reply]
> -- create your thread
> script.register_looped("autoheal", function(script)
>   script:yield() -- yield the execution until the next frame

That's not required. The function already yields automatically:
https://github.com/YimMenu/YimMenu/blob/6bc6801fec6ceadb9551cd4baeb9521078b65e8c/src/lua/bindings/script.cpp#L75-L82
